apollo-router/src/plugins/authentication/subgraph.rs

```diff
@@ -192,12 +198,100 @@ pub(crate) struct SigningParams {
 
 #[derive(Clone)]
 pub(crate) struct SigningParamsConfig {
-    credentials_provider: Arc<dyn ProvideCredentials>,
+    credentials_provider: CredentialsProvider,
     region: Region,
     service_name: String,
     subgraph_name: String,
 }
 
+#[derive(Clone, Debug)]
+struct CredentialsProvider {
+    credentials: Arc<RwLock<Credentials>>,
+    _credentials_updater_handle: Arc<JoinHandle<()>>,
+    #[allow(dead_code)]
+    refresh_credentials: Sender<()>,
+}
+
+// Refresh token if it will expire within the next 5 minutes
+const MIN_REMAINING_DURATION: Duration = std::time::Duration::from_secs(60 * 5);
+
+impl CredentialsProvider {
+    async fn from_provide_credentials(
+        provide_credentials: impl ProvideCredentials + 'static,
+    ) -> Result<Self, CredentialsError> {
+        let credentials_provider = SharedCredentialsProvider::new(provide_credentials);
+        let (sender, mut refresh_credentials_receiver) = tokio::sync::mpsc::channel(1);
+        let credentials = credentials_provider.provide_credentials().await?;
+        let mut refresh_timer = next_refresh_timer(&credentials);
+        let credentials = Arc::new(RwLock::new(
+            credentials_provider.provide_credentials().await?,
+        ));
+        let c2 = credentials.clone();
+        let handle = tokio::spawn(async move {
+            loop {
+                tokio::select! {
+                    _ = tokio::time::sleep(refresh_timer) => {
+                        // Refresh credentials
+                        if let Ok(new_credentials) = credentials_provider.provide_credentials().await {
+                            refresh_timer = next_refresh_timer(&new_credentials);
```

Comment:
`refresh_timer` is a `Duration`, which means that on `Err` we will wait again for `X`, `X` potentially being greater than the validity period of the credentials. Should we handle `Err`s here so we have a shorter refresh timer window?

Reply:
Great idea, i went and used a 1 minute retry timer, we'll see how it goes!